fixes for CONFIG_BOOT_TIME_MEASUREMENT

gnuless opened the pull request:
The boot time measurement sample was giving bogus values on x86: an
assumption was made that the system timer is in sync with the CPU TSC,
which is not the case on most x86 boards.

Boot time measurements are no longer permitted unless the timer source
is the local APIC. To avoid issues of TSC scaling, the startup datum
has been forced to 0, which is in line with the ARM implementation
(which is the only other platform which supports this feature).

Cleanups along the way:

As the datum is now assumed zero, some variables are removed and
calculations simplified. The global variables involved in boot time
measurements are moved to the kernel.h header rather than being
redeclared in every place they are referenced. Since none of the
measurements actually use 64-bit precision, the samples are reduced
to 32-bit quantities.

In addition, this feature has been enabled in long mode.

Fixes: #19144

Commit:
subsys/debug: eliminate PEFORMANCE_METRICS and CPU_CLOCK_FREQ_MHZ

Neither of these options is actually used; CPU_CLOCK_FREQ_MHZ appears
to have been part of x86 BOOT_TIME_MEASUREMENT at some point, and
PERFORMANCE_METRICS is the stillborn cousin of EXECUTION_BENCHMARKS.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
tests/benchmarks: fix BOOT_TIME_MEASUREMENT

The boot time measurement sample was giving bogus values on x86: an
assumption was made that the system timer is in sync with the CPU TSC,
which is not the case on most x86 boards.

Boot time measurements are no longer permitted unless the timer source
is the local APIC. To avoid issues of TSC scaling, the startup datum
has been forced to 0, which is in line with the ARM implementation
(which is the only other platform which supports this feature).

Cleanups along the way:

As the datum is now assumed zero, some variables are removed and
calculations simplified. The global variables involved in boot time
measurements are moved to the kernel.h header rather than being
redeclared in every place they are referenced. Since none of the
measurements actually use 64-bit precision, the samples are reduced
to 32-bit quantities.

In addition, this feature has been enabled in long mode.

Fixes: #19144

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: refactor APIC timer configuration to SoC level

The APIC is part of the SoC, not the board, so move the defaults down.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
tests/benchmarks/boot_time: disable on QEMU targets

These all currently use the HPET timer, and thus can't build with
BOOT_TIME_MEASUREMENT enabled, so disable for now. This test is
basically a build-only test, so we're covered with the other x86
targets (gpmrb, up_squared) for now.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -31,8 +31,16 @@ __start:
+#ifdef CONFIG_BOOT_TIME_MEASUREMENT
+	xorl	%eax, %eax
+	xorl	%edx, %edx
+	movl	$X86_TIME_STAMP_COUNTER_MSR, %ecx
```
indentation looks off here

gnuless replied:
Whoops, I hastily airlifted this from the ia32 version. Fixed the spacing between the mnemonic and operands.

andrewboie commented on tests/benchmarks/boot_time/src/main.c:
```diff
@@ -45,37 +35,29 @@ void main(void)
 	TC_PRINT("_start->main(): %u cycles, %u us\n",
-		 (u32_t)(s_main_time_stamp & 0xFFFFFFFFULL),
-		 (u32_t)  (main_us  & 0xFFFFFFFFULL));
+		 __main_time_stamp,
```
here as well

gnuless replied:
Not sure what you were seeing here -- if it was an artifact of github (I've seen that before) or a genuine ugly -- but I've reformatted with a slightly different style.

tyrantlizard77 reviewed (commented):
assuming GH is rendering properly I'm seeing some tabs/spaces issues but otherwise LGTM

Commit:
tests/benchmarks/boot_time: disable on inapplicable targets

Minnowboard and ACRN do not use the new APIC timer, so the
benchmark code will not run on them (yet).

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
boards/x86/gpmrb: disable Multiboot header generation

We don't require this to boot with SBL, so disable for now.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

nashif commented on include/kernel.h:
```diff
@@ -23,6 +23,11 @@
 extern "C" {
 #endif
 
+#ifdef CONFIG_BOOT_TIME_MEASUREMENT
```
None of this is public API material, can we put this somewhere else other than kernel.h which is a public API header?

gnuless replied:
I'm not entirely sure where to put it.  These variables are defined and manipulated from `kernel/`, and used in `tests/` (which probably should be `samples/`, how is this a test?), and are enabled by an option in `subsys/debug/`.  Whoever originally put this in seemed confused about its categorization, as well.

Previously these variables were simply redeclared in all these places, but I don't think going back to that is a good solution.

I'm open to suggestions.

gnuless replied:
How about `kernel/include/kernel_internal.h` ?

andrewboie replied:
kernel_internal.h seems a good spot.

andrewboie replied:
I'm inclined to merge this as-is, I'm cleaning up how the arch interfaces are specified anyway and will take this under consideration.

ioannisg reviewed (commented):
The ARM changes is just the clean-up, right?

gnuless commented:
> The ARM changes is just the clean-up, right?

Yeah, I removed a variable (since it always had the same value everywhere), so I removed the initialization in the ARM code, too.
